spqr/profile_spqr.py

```diff
@@ -0,0 +1,238 @@
+        spqr_cuda.spqr_mul_timer(
+            spqr_device.m,
+            spqr_device.n,
+            spqr_device.bits,
+            spqr_device.beta1,
+            spqr_device.beta2,
+            spqr_device.dense_weights,
+            spqr_device.row_offsets,
+            spqr_device.col_vals,
+            spqr_device.nnz,
+            x,
+            y,
+            runs[i],
+            feature_flag)
+
+    return y, runs
+
+
+def torch_mul_timer(deq_w, x, num_runs):
+    if len(deq_w.shape) == 1:
+        n = x.shape[0]
+        m = deq_w.shape[0] // n
+    else:
+        m, n = deq_w.shape
+
+    assert (n == x.shape[0])
+
+    runs = torch.empty(num_runs).cpu().float()
+
+    y = torch.zeros(m, dtype=x.dtype, device=x.device)
+
+    for i in range(num_runs):
+        y = torch.zeros_like(y)
+        spqr_cuda.torch_mul_fp16(m, n, deq_w, x, y, runs[i])
+
+    return y, runs
+
+
+if __name__ == '__main__':
+    torch_runs = {}
+
+    parser = argparse.ArgumentParser(add_help=True)
+
+    parser.add_argument(
+        "--tensor_path",
+        type=str,
+        required=True,
+        help="Path to folder containing the tensors of the form"
+             "model_path/"
+             "   0/"
+             "       tensor0"
+             "       tensor1"
+    )
+
+    parser.add_argument(
+        "--ptcsr_path",
+        type=str,
+        required=False,
+        help="Path to folder containing the tensors of the form"
+             "model_path/"
+             "   0/"
+             "       tensor0"
+             "       tensor1"
+    )
+
+
+    parser.add_argument(
+        "--output_path",
+        type=str,
+        help="Path to results *.csv file.",
+    )
+
+    args = parser.parse_args()
+
+    with open(args.output_path, 'w') as f:
+        run_ptcsr = args.ptcsr_path is not None
+
+        base_path = args.tensor_path
+        base_path_modified_csr = args.ptcsr_path
+
+        seed = 1
+        np.random.seed(seed)
+        torch.random.manual_seed(seed)
+
+        NUM_RUNS = 2
+        WARMUP = 1
+
+        device = torch.device('cuda')
+
+        for m in [4096, 11008]:
+            for n in [4096, 11008]:
+                d = torch.zeros((m, n), dtype=torch.float16, device=device)
+                x = torch.zeros(n, dtype=torch.float16, device=device)
+                y, dense_runs = torch_mul_timer(d, x, NUM_RUNS)
+                this_algorithm = dense_runs[WARMUP:].min()
+                torch_runs[(m, n)] = this_algorithm
+                torch.cuda.empty_cache()
+                time.sleep(2)
+
+        csr_folders = os.listdir(base_path)
+
+
+        if run_ptcsr:
+            folders_modified_csr = os.listdir(base_path_modified_csr)
+        else:
+            folders_modified_csr = os.listdir(base_path)
+
+        csr_folders.sort()
+        folders_modified_csr.sort()
+
+        methods = [
+            inference.FeatureFlags.SPARSE_FUSED_FP32,
+        ]
+
+        f.write('Layer;Tensor Name;M;N;Sparsity (%)')
```

Comment:
M:N Sparsity? Do you mean the sparsity matrix % of sparsity?

Reply:
M,N are csv table column names. :)